patched-vscode/extensions/sagemaker-terminal-crash-mitigation/src/extension.ts

```diff
@@ -0,0 +1,103 @@
+import * as vscode from 'vscode';
+import { exec } from 'child_process';
+
+const logPrefix = '[sagemaker-terminal-crash-mitigation]';
+
+export function activate(_context: vscode.ExtensionContext) {
+    let lastTerminal: vscode.Terminal | undefined;
+    let lastProcessId: number | undefined;
+    let lastOpenedTime: number | undefined;
+
+     /**
+     * Event listener for when a new terminal is opened.
+     * Tracks the terminal's process ID and the time it was opened.
+     */
+    vscode.window.onDidOpenTerminal(async terminal => {
+        lastTerminal = terminal;
+        lastOpenedTime = Date.now();
+        try {
+            lastProcessId = await terminal.processId;
+            console.log(`${logPrefix} Terminal opened: PID ${lastProcessId}, Time: ${lastOpenedTime}`);
+        } catch (error) {
+            console.error(`${logPrefix} Error getting process ID: ${error}`);
+        }
+    });
+
+    /**
+     * Event listener for when a terminal is closed.
+     * Checks if the closed terminal is the one that was last opened,
+     * and if it closed within 1 second. If no other terminals are active,
+     * executes a command to kill all bash processes.
+     */
+    vscode.window.onDidCloseTerminal(async terminal => {
+        if (lastTerminal && lastProcessId && lastOpenedTime) {
+            try {
+                const currentProcessId = await terminal.processId;
+                console.log(`${logPrefix} Terminal closed: PID ${currentProcessId}`);
+
+                if (currentProcessId === lastProcessId) {
+                    const timeElapsed = Date.now() - lastOpenedTime;
+                    console.log(`${logPrefix} Time elapsed since opening: ${timeElapsed}ms`);
+
+                    if (timeElapsed < 1000) {
+                        const remainingTerminals = vscode.window.terminals.length;
+                        console.log(`${logPrefix} Number of remaining terminals: ${remainingTerminals}`);
+
+                        if (remainingTerminals === 0) {
+                            console.log(`${logPrefix} No other active terminals. Executing kill command.`);
+                            execKillCommand();
```

Comment:
As discussed, we can also consider automatically retry so customers don't have to manually retry opening terminal after we kill all bashes - but not blocking